00-programming-fundamentals/debugging-user-reports.md

```diff
@@ -0,0 +1,43 @@
+
+## Example
+Here's an example of the kind of user report that you might receive, and the resulting repro steps we could generate. For this example imagine that you are working on an e-commerce website that has a shopping cart feature.
+
+### User Report
+> I added the Silver Tails Dog Infrared Massager. But I added two somehow. When I tried to remove one from my cart it went away entirely!
+
+### Repro Steps
+1. Open the website in Chrome.
+1. Sign up for a new account using the standard sign up procedure.
+1. Use the product search bar to search for "silver tails dog infrared massager"
+1. Click on the result labeled "Silver Tails Dog Infrared Massager with 2-Heated Heads."
+1. Click on the "Add to Cart" button.
+1. Press the browser's back button.
+1. Click on the "Add to Cart" button again.
+1. Observe that the quantity for the above item is listed as 2.
+1. Open the quantity drop-down for the item and select "1".
+1. Click the "Update item" button next to the drop-down.
+1. Observe that the item "Silver Tails Dog Infrared Massager with 2-Heated Heads." is no longer listed in the cart.
+
+### Important Considerations
+As you can see above the repro steps for this example bug are much more detailed and thorough than what the user presented. This additional detail will be instrumental in helping a developer (possibly you) pin-point the exact code that was involved when the bug was triggered. There are some additional things to take note of:
+
+* The repro steps should be as exact as possible. For the first step we didn't say "Open the website in a browser" -- this is important to minimize the variation between how you or the QA person reproduced the bug and how the developer fixing it is reproducing the bug.
+
+  Sometimes if a developer receives an issue ticket (how bugs and feature work are tracked) with some lenience or variation in how the repro steps could be followed they will try it one way and not trigger the bug, and then they will mark the issue as resolved because it's not a real bug.
+* The repro steps should involve the original report's details to the greatest extent possible. In these repro steps we continued to use the dog massager item, even though this bug might affect other items.
+
+  While you can sometimes generalize the repro steps, it's not always obvious when that is possible or not. In those cases err on the side of caution and re-use the details from the initial report -- they might be critical to diagnosing the underlying bug.
+* The repro steps should conclude with an observation. Generally the final part of any set of repro steps is to observe the unexpected behavior, this makes it clear specifically how the bug is manifesting (at a user level) in the final context of the repro steps.
```

Comment:
Adding a vocab list here would be good. Things like:

```
### Vocab
- **User report:** A description of a bug found by a user. Often described in terms of the UI, since the user doesn't have any insight into how your app actually works.
- **Repro:** A set of steps to consistently reproduce a bug.
- **Observation:** The last step of a repro. Includes a description of what was expected and what actually happened.
```